accepted/2023/net8.0-browser-tfm.md

```diff
@@ -0,0 +1,187 @@
+
+* Projects can have specific dependencies for the browser (like MAUI)
+  ```xml
+  <ItemGroup Condition="'$(Browser)' == 'true'">
+    <Reference Include="Microsoft.AspNetCore.Components.WebAssembly" />
+    <Reference Include="System.Net.Http.Json" />
+  </ItemGroup>
+  ```
+* Project can have different API area surfaces for the browser
+  ```csharp
+  public class RecipesStore
+  {
+  #if BROWSER
+      private readonly HttpClient _http;
+  #else
+      IDictionary<string, Recipe> recipes;
+      ConcurrentDictionary<string, byte[]> images = new();
+      InMemorySearchProvider searchProvider;
+  #endif
+
+  #if !BROWSER
+      public RecipesStore()
+      {
+        ...
+      }
+  #else
+      public RecipesStore(HttpClient http)
+      {
+          _http = http;
+      }
+  #endif
+
+  #if BROWSER
+      public async Task<string> AddImage(Stream imageData)
+      {
+          var response = await _http.PostAsync("api/images", new StreamContent(imageData));
+          return await response.Content.ReadAsStringAsync();
+      }
+
+      public async Task<string> AddRecipe(Recipe recipe)
+      {
+          var response = await _http.PostAsJsonAsync("api/recipes", recipe);
+          return await response.Content.ReadAsStringAsync();
+      }
+  #else
+      public async Task<IEnumerable<Recipe>> GetRecipes(string? query)
+      {
+          ...
+      }
+
+      public Task<Recipe?> GetRecipe(string id)
+      {
+          ...
+      }
+
+      public Task<Recipe> UpdateRecipe(Recipe recipe)
+      {
+          ...
+      }
+
+      public Task<string> AddRecipe(Recipe recipe)
+      {
+          ...
+      }
+
+      public async Task<string> AddImage(Stream imageData)
+      {
+          ...
+      }
+
+      public Task<byte[]> GetImage(string filename)
+          => ...;
+  #endif
+  }
+  ```
+
+## Requirements
+
+Introduce a new net8.0-browser TFM that enables developers to author dlls with an API surface tailored for running in the browser, to leverage browser specific APIs, and to pivot their dependecies based on the platform.
+
+The net8.0-browser TFM will be a superset of net8.0, which means that an app or library targeting net8.0-browser will be able to consume net8.0 dependencies (for unsupported APIs we will continue using `[SupportedOSPlatform]`), however if a net8.0-browser version of the library is available, that would be preferred over the net8.0 version.
+
+The general rule will be that this TFM will work in the same way as net8.0-android, net8.0-ios, net8.0-windows, net8.0-maccatalyst, etc. do for Maui apps today.
+
+### Goals
+
+* Enable single project experiences for Blazor Webassembly applications that target ASP.NET Core environments and the browser with the same codebase.
+* Enable library authors to have dependencies thar are tailored for the browser.
+
+### Non-Goals
+
+**TL;DR**
+* Per browser TFM
+* Per browser version TFM
+* Browser specific API sets (although might be expanded in the future to cover this)
```

Comment:
I guess NodeJS is just another "browser" flavor worth mentioning ?

Reply:
I would not bundle node with other browsers, it has a different set of APIs and capabilities, we should handle that separately in the future

Reply:
That's not how runtime treats it today. Do we need yet another TFM for it ?
@lewing what's your opinion ?

Reply:
@pavelsavara but you do not offer different APIs tailored for node either, isn't it?

I would be conservative here about introducing a TFM for the same reasons we were before in the initial wasm versions. We can always add it later, but we can't take it back.

I would not expect us to have any sort of relationship between `net8.0-browser` and `net8.0-(xxx)` even if some APIs might be similar (like fetch). As they are in general different environments and evolve at different speeds for different needs.

It would come down to how many users you have that want to target node, and potentially also provide a different API shape and dependencies for the browser.

This proposal does not change what the runtime will do today. The runtime can continue shipping as net8.0 (as it does today for Maui) but the TFM gives the ecosystem the ability to choose dependencies, API shapes, and provide different implementations in a natural way. 

Apps can target net8.0-browser and get access to browser specific APIs and implementations provided by packages (similar to Maui).

If later on we want to provide a set of out of the box APIs, that becomes a different story.

Reply:
We do not offer browser-only features either. And that's how it is the same API surface with NodeJS too.